api/cpp/include/slint.h

```diff
@@ -1049,6 +1049,179 @@ class MapModel : public Model<MappedModelData>
     std::function<MappedModelData(const SourceModelData &)> map_fn;
 };
 
+template<typename ModelData>
+class SortModel;
+
+namespace private_api {
+template<typename ModelData>
+struct SortModelInner : private_api::ModelChangeListener
+{
+    SortModelInner(std::shared_ptr<slint::Model<ModelData>> source_model,
+                   std::function<bool(const ModelData &, const ModelData &)> sort_fn,
+                   slint::SortModel<ModelData> &target_model)
+        : source_model(source_model), sort_fn(sort_fn), target_model(target_model)
+    {
+    }
+
+    void row_added(int first_inserted_row, int count) override
+    {
+        if (sorted_rows_dirty) {
+            reset();
+            return;
+        }
+
+        // Adjust the existing sorted row indices to match the updated source model
+        for (auto it = sorted_rows.begin(), end = sorted_rows.end(); it != end; ++it) {
+            if (*it >= first_inserted_row)
+                *it += count;
+        }
+
+        for (int row = first_inserted_row; row < first_inserted_row + count; ++row) {
+
+            ModelData inserted_value = *source_model->row_data(row);
+            auto insertion_point =
+                    std::lower_bound(sorted_rows.begin(), sorted_rows.end(), inserted_value,
+                                     [this](int sorted_row, const ModelData &inserted_value) {
+                                         auto sorted_elem = source_model->row_data(sorted_row);
+                                         return sort_fn(*sorted_elem, inserted_value);
+                                     });
+
+            insertion_point = sorted_rows.insert(insertion_point, row);
+            target_model.row_added(insertion_point - sorted_rows.begin(), 1);
+        }
+    }
+    void row_changed(int changed_row) override
+    {
+        if (sorted_rows_dirty) {
+            reset();
+            return;
+        }
+
+        for (auto it = sorted_rows.begin(); it != sorted_rows.end(); ++it) {
+            if (*it == changed_row) {
+                it = sorted_rows.erase(it);
+                target_model.row_removed(it - sorted_rows.begin(), 1);
+                break;
+            }
+        }
+
+        ModelData changed_value = *source_model->row_data(changed_row);
+        auto insertion_point =
+                std::lower_bound(sorted_rows.begin(), sorted_rows.end(), changed_value,
+                                 [this](int sorted_row, const ModelData &changed_value) {
+                                     auto sorted_elem = source_model->row_data(sorted_row);
+                                     return sort_fn(*sorted_elem, changed_value);
+                                 });
+
+        insertion_point = sorted_rows.insert(insertion_point, changed_row);
+        target_model.row_added(insertion_point - sorted_rows.begin(), 1);
```

Comment:
This will always turn a changed into a remove/add sequence, even when that might not be necessary as the sorted position does not actually change.

Can that be a problem later on? I doubt it.

Reply:
I think you're right, I should add code to detect the situation where the insertion point is the same as the removed row. And if that's the same, just emit row_changed.